src/twisted/application/test/test_internet.py

```diff
@@ -771,6 +789,25 @@ def test_clientConnectionLost(self):
             self.successResultOf(service.whenConnected()), cq.applicationProtocols[1]
         )
 
+    def test_clientConnectionRefused(self) -> None:
+        """
+        When the client factory refuses connections by returning None from
+        buildProtocol, the factory keeps attempting new connections.
+        """
+        clock = Clock()
+        cq, service = self.makeReconnector(
+            clock=clock, fireImmediately=False, refuseConnections=True
+        )
+        self.assertEqual(len(cq.connectQueue), 1)
```

Comment:
I find it hard to read this test... as not being familiar with the code.

I can see where this happends in the test code

Maybe add some comments to explain the steps like.

```python
# As start, there is connection made.
self.assertEqual(len(cq.connectQueue), 1)
# The service refuses the new connection.
# This is signaled by returning C{None}
cq.connectQueue[0].callback(None)
# No new connections are triggered at this point.
self.assertEqual(len(cq.connectQueue), 1)
# Once connected the service will be notified on the new connection
whenConnected = service.whenConnected()
self.assertNoResult(whenConnected)
# Once the servic had handled the new connection event,
# and the handling fails, it will try a new connection.
whenConnected.addErrback(lambda ignored: ignored.trap(CancelledError))
# After the retry period, a new connection is made.
clock.advance(AT_LEAST_ONE_ATTEMPT)
self.assertEqual(len(cq.connectQueue), 2)
# I have no idea why we need this call.
cq.connectQueue[1].callback(None)
```